fuel-txpool/src/containers/dependency.rs

```diff
@@ -0,0 +1,635 @@
+                }
+            }
+        }
+        Ok((max_depth, db_coins, db_contracts, collided))
+    }
+    /// insert tx inside dependency
+    /// return list of transactions that are removed from txpool
+    pub async fn insert<'a>(
+        &'a mut self,
+        txs: &'a HashMap<TxId, ArcTx>,
+        db: &dyn TxPoolDB,
+        tx: &'a ArcTx,
+    ) -> anyhow::Result<Vec<ArcTx>> {
+        let (max_depth, db_coins, db_contracts, collided) = self.check_for_colision(txs, db, tx)?;
+
+        // now we are sure that transaction can be included. remove all collided transactions
+        let mut removed_tx = Vec::new();
+        for collided in collided.into_iter() {
+            let collided = txs
+                .get(&collided)
+                .expect("Collided should be present in txpool");
+            removed_tx.extend(self.recursively_remove_all_dependencies(txs, collided.clone()));
+        }
+
+        // iterate over all inputs and spend parent coins/contracts
+        for input in tx.inputs() {
+            match input {
+                Input::Coin { utxo_id, .. } => {
+                    // spend coin
+                    if let Some(state) = self.coins.get_mut(utxo_id) {
+                        state.is_spend_by = Some(tx.id());
+                    }
+                }
+                Input::Contract { contract_id, .. } => {
+                    // contract
+                    if let Some(state) = self.contracts.get_mut(contract_id) {
+                        state.used_by.insert(tx.id());
+                    }
+                }
+            }
+        }
+
+        // insert all coins/contracts that we got from db;
+        self.coins.extend(db_coins.into_iter());
+        self.contracts.extend(db_contracts.into_iter());
+
+        // iterate over all outputs and insert them, marking them as available.
+        for (index, output) in tx.outputs().iter().enumerate() {
+            let is_contract = match output {
+                Output::Coin { .. }
+                | Output::Withdrawal { .. }
+                | Output::Change { .. }
+                | Output::Variable { .. } => None,
+                Output::Contract { input_index, .. } => {
+                    if let Input::Contract { contract_id, .. } = tx.inputs()[*input_index as usize]
+                    {
+                        Some(contract_id)
+                    } else {
+                        panic!("InputIndex for contract should be always correct");
+                    }
+                }
+                Output::ContractCreated { contract_id } => Some(*contract_id),
+            };
+            let utxo_id = UtxoId::new(tx.id(), index as u8);
+            if let Some(contract_id) = is_contract {
+                // insert contract
+                self.contracts.insert(
+                    contract_id,
+                    ContractState {
+                        depth: max_depth,
+                        used_by: HashSet::new(),
```

Comment:
Wouldn't this overwrite all the used_by tx ID set on line 307?:
```
                        // insert into contract
                        db_contracts
                            .entry(*contract_id)
                            .or_insert(ContractState {
                                used_by: HashSet::new(),
                                depth: 0,
                                origin: None, //there is no need for owner if contract is in db
                                gas_price: GasPrice::MAX,
                            })
                            .used_by
                            .insert(tx.id());
```

Reply:
Yep, this is a bug, if output is contract it should do nothing, When i wrote this one month ago i didnt fully grasp all outputs, for withdrawal output it is same story.